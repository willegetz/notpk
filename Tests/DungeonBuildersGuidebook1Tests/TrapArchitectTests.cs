using System.Collections.Generic;
using System.Linq;
using System.Text;
using ApprovalTests;
using ApprovalTests.Reporters;
using System.Xml.Linq;
using DungeonBuildersGuidebook1;
using DungeonBuildersGuidebook1.TrapComponentObjects;
using RpgTools.Dice;
using System;
using NUnit.Framework;

namespace DungeonBuildersGuidebook1Tests
{
	[TestFixture]
	[UseReporter(typeof(DiffReporter))]
	public class TrapArchitectTests
	{
		// The user should also be allowed to pick their Trap Basic
		//		Some form of display to pick from

		[Test]
		[Ignore("Completely Random, different results each time.")]
		public void GenerateRandomTrapsTest()
		{
			var architect = new TrapArchitect();
			var traps = new List<TheTrap>();

			var numberOfTraps = 100;
			for (int i = 1; i <= numberOfTraps; i++)
			{
				var theTrap = new TheTrap();

				theTrap.SetTrapBase(architect.GetRandomTrapBase());
				theTrap.SetTrapEffect(architect.GetRandomTrapEffect());
				theTrap.SetTrapDamage(architect.GetRandomTrapDamage());
				traps.Add(theTrap);
			}

			var sb = new StringBuilder();
			foreach (var trap in traps)
			{
				sb.AppendLine(trap.ToString());
				sb.AppendLine();
			}
			Approvals.VerifyAll(traps, "Traps");
		}

		[Test]
		public void GetSpecificTrapsTest()
		{
			var architect = new TrapArchitect();
			var traps = new List<TheTrap>();
			
			int tableMax = 100;

			var trapDiceRolls = new List<int>();
			for (int i = 1; i < tableMax; i += 2)
			{
				trapDiceRolls.Add(i);
			}

			var damageDiceRolls = new List<int>();
			var damageResult = 3;
			for (int i = 0; i < trapDiceRolls.Count; i++)
			{
				if (damageResult <= 18)
				{
					damageDiceRolls.Add(damageResult);
					damageResult++;
				}
				else
				{
					damageResult = 3;
					damageDiceRolls.Add(damageResult);
					damageResult++;
				}
			}

			for (int i = 0; i < trapDiceRolls.Count; i++)
			{
				var theTrap = new TheTrap();
				theTrap.SetTrapBase(architect.GetSpecificTrapBase(trapDiceRolls[i]));
				theTrap.ChangeMechanismType("magical");
				theTrap.SetTrapEffect(architect.GetSpecificTrapEffect(trapDiceRolls[i]));
				theTrap.SetTrapDamage(architect.GetSpecificTrapDamage(damageDiceRolls[i]));
				traps.Add(theTrap);
			}

			traps[2].ChangeAnEffect(1, architect.GetSpecificTrapEffect(20));
			traps[23].ChangeAnEffect(1, architect.GetSpecificTrapEffect(30));
			traps[49].ChangeAnEffect(1, architect.GetSpecificTrapEffect(40));
			traps[49].ChangeAnEffect(2, architect.GetSpecificTrapEffect(50));

			Approvals.VerifyAll(traps, "Traps");
		}

		[Test]
		public void RerollTrapTest()
		{
			var architect = new TrapArchitect();
			var trap = new TheTrap();
			var sb = new StringBuilder();

			int firstTrapBase = 2;
			int firstTrapEffect = 24;
			int firstDamageEffect = 10;

			trap.SetTrapBase(architect.GetSpecificTrapBase(firstTrapBase));
			trap.SetTrapEffect(architect.GetSpecificTrapEffect(firstTrapEffect));
			trap.SetTrapDamage(architect.GetSpecificTrapDamage(firstDamageEffect));

			sb.AppendLine("Initial Trap");
			sb.AppendLine(trap.ToString());
			sb.AppendLine();

			int rerollTrapBase = 67;
			int rerollTrapEffect = 84;
			int indexOfEffectToChange = 0;
			trap.SetTrapBase(architect.GetSpecificTrapBase(rerollTrapBase));
			trap.ChangeAnEffect(indexOfEffectToChange, architect.GetSpecificTrapEffect(rerollTrapEffect));
			
			sb.AppendLine("New Trap Base");
			sb.AppendLine(trap.ToString());
			
			Approvals.Verify(sb.ToString());
		}

		[Test]
		public void GasSubtableTest()
		{
			var architect = new TrapArchitect();
			var trap = new TheTrap();

			int trapBaseRoll = 25;
			int trapEffectRoll = 38;
			int trapDamageRoll = 6;
			int effectIndex = 0;
			int gasTypeRoll = 3;

			trap.SetTrapBase(architect.GetSpecificTrapBase(trapBaseRoll));
			trap.SetTrapEffect(architect.GetSpecificTrapEffect(trapEffectRoll));
			trap.ChangeAnEffectSubtableDescription(effectIndex, architect.GetSpecificSubtableEffect(trap.GetSubtableType(effectIndex), gasTypeRoll));
			trap.SetTrapDamage(architect.GetSpecificTrapDamage(trapDamageRoll));

			Approvals.Verify(trap.ToString());
		}

		[Test]
		public void PitContentsSubtableTest()
		{
			var architect = new TrapArchitect();
			var trap = new TheTrap();

			int trapBaseRoll = 84;
			int TrapEffectRoll = 94;
			int trapDamageRoll = 15;
			int effectIndex = 0;
			int pitContentsRoll = 5;

			trap.SetTrapBase(architect.GetSpecificTrapBase(trapBaseRoll));
			trap.SetTrapEffect(architect.GetSpecificTrapEffect(TrapEffectRoll));
			trap.ChangeAnEffectSubtableDescription(effectIndex, architect.GetSpecificSubtableEffect(trap.GetSubtableType(effectIndex), pitContentsRoll));
			trap.SetTrapDamage(architect.GetSpecificTrapDamage(trapDamageRoll));

			Approvals.Verify(trap.ToString());
		}

		[Test]
		public void TrapEffectFactoryTest()
		{
			var architect = new TrapArchitect();
			var factoryEffect1 = architect.GetSpecificTrapEffect1(99);
			var factoryEffect2 = architect.GetSpecificTrapEffect1(100);

			var logicEffect1 = architect.GetSpecificTrapEffect(1).First().EffectDescription.ToLower();
			var logicEffect2 = architect.GetSpecificTrapEffect(2).First().EffectDescription.ToLower();

			Assert.AreEqual(factoryEffect1, logicEffect1);
			Assert.AreEqual(factoryEffect2, logicEffect2);
		}

		[Test]
		public void RandomEntryTest()
		{
			DiceCup.SetRandom(new Random(4));
			var architect = new TrapArchitect();

			var factoryList = new List<string>();
			var firstList = new List<string>();

			int repeat = 1000;

			// roll 455 gave an interesting result
			for (int i = 1; i <= repeat; i++)
			{
				var result = DiceCup.Roll("1d100");
				factoryList.Add(architect.GetSpecificTrapEffect1(result));
				
				//firstList.Add(architect.GetSpecificTrapEffect(result).First().EffectDescription.ToLower());
			}

			Approvals.VerifyAll(factoryList, "List");
			//Approvals.VerifyAll(firstList, "List");

		}

		[Test]
		//[Ignore("Completely random")]
		public void RandomFactoryTest()
		{
			var architect = new TrapArchitect();
			var effectList = new List<string>();
			int repeat = 20000;
			for (int i = 0; i < repeat; i++)
			{
				effectList.Add(architect.GetTrapEffectFactory());
			}

			Approvals.VerifyAll(effectList, "Traps");

		}

		[Test]
		public void RandomTrapBaseFactoryTest()
		{
			var architect = new TrapArchitect();
			var trapBase = architect.GetTrapBaseFactory();
			Approvals.Verify(trapBase);
		}

		[Test]
		public void RandomTrapConstructionTest()
		{
			var architect = new TrapArchitect();
			var traps = new List<StringBuilder>();
			var iterator = 1000;
			for (int i = 0; i < iterator; i++)
			{
				var sb = new StringBuilder();
				sb.AppendLine(architect.GetTrapBaseFactory());
				sb.AppendLine(architect.GetTrapEffectFactory());
				traps.Add(sb);
			}

			Approvals.VerifyAll(traps, "Traps");
		}

		[Test]
		[Ignore("Experiment Workspace")]
		public void TrapEffectsTest()
		{
			// A trap effect has at minimum an effect name and an upper bound on the range
			// A trap effect may have additional rolls 
			//	These rolls will be for random values defined by the trap effect
			// A trap effect may require the consultation of another table
			//	These consultations will be performed on tables that are named
			//		These tables handle their own information
			var filePath = @"..\..\..\..\DungeonBuildersGuidebook1\DataFiles\TrapEffectsAndTraits.xml";
			var root = XElement.Load(filePath);

			var effects = root.Descendants("Effect").Select(p => new TrapEffects()
																	{
																		RollUpperBound = int.Parse(p.Element("RollUpperBound").Value),
																		EffectDescription = p.Element("EffectDescription").Value,
																		//EffectRollRequired = bool.Parse(p.Element("EffectRollRequired").Value),
																		//EffectDiceToRoll = p.Elements("EffectDiceToRoll").Select(e => e.Value).ToList(),
																	}
															);

			var sb = new StringBuilder();
			var results = new List<int>();
			//foreach (var item in effects.Where(r => r.RollAgain))
			//{
			//    foreach (var die in item.EffectDiceToRoll)
			//    {
			//        var bag = new DiceCup(DiceDefinition.Parse(die));
			//        results.Add(bag.Roll());
			//    }

			//    var blah = string.Format(item.EffectDescription, results[0]);

			//    item.EffectDescription = blah;
			//}

			foreach (var item in effects)
			{
				sb.AppendLine(item.EffectDescription);
			}
			Approvals.Verify(sb.ToString());
		}

		[Test]
		[Ignore] // Playgrounds
		public void TestLoadXmlElement()
		{
			var trapArchitect = new TrapArchitect();
			var filePath = @"..\..\..\data.xml";

			var root = XElement.Load(filePath);
			var blah = root.Descendants("Data").Select(p => new FakeProduct()
												{
													Category = p.Element("Category").Value,
													Quantity = int.Parse(p.Element("Quantity").Value),
													Price = double.Parse(p.Element("Price").Value),
												}
									);

			var blah3 = blah.Where(c => c.Category == "A");


			var sb = new StringBuilder();
			foreach (var item in blah)
			{
				sb.AppendLine(item.ToString());
			}

			sb.AppendLine();

			sb.AppendLine("The items from Category A");
			foreach (var item in blah3)
			{
				sb.AppendLine(item.ToString());
			}
			
			IEnumerable<decimal> prices1 =
				from el in root.Elements("Data")
				let price = (decimal)el.Element("Price")
				orderby price
				select price;
			foreach (decimal el in prices1) ;

			Approvals.Verify(sb.ToString());
		}
	}

	class FakeTrapArchitect : TrapArchitect
	{
		public TrapBases FakeGetRandomTrapBase(int objectIndex)
		{
			return base.trapBaseLogic.GetRandomTrapBase();
		}

		public string FakegetMechanismType(TrapBases trapBase, int result)
		{
			if (trapBase.MechanismTypeSpecified)
			{
				return result > 50 ? "magical" : "mechanical";
			}
			else
			{
				return string.Empty;
			}
		}
	}
}
