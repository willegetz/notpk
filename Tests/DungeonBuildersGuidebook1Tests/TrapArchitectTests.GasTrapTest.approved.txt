Gas Trap![0] = The vegetation
Releases a gas: Obscuring
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![1] = The hook
Releases a gas: Weakening
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![2] = The organ
Releases a gas: Sleep Inducing
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![3] = The column
Releases a gas: Blinding
Infliciting 4d6 points of damage in addition to the other effects listed

Gas Trap![4] = The chair
Releases a gas: Obscuring
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![5] = The bed
Releases a gas: Sleep Inducing
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![6] = The rug
Releases a gas: Corroding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![7] = The forcefield
Releases a gas: Corroding
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![8] = The door
Releases a gas: Poisonous
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![9] = The bed
Releases a gas: Nauseating
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![10] = The holy/unholy font
Releases a gas: Blinding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![11] = The floor
Releases a gas: Blinding
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![12] = The pendulum
Releases a gas: Nauseating
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![13] = The hook
Releases a gas: Nauseating
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![14] = The barrel
Releases a gas: Weakening
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![15] = The manacles
Releases a gas: Terrifying
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![16] = The painting
Releases a gas: Obscuring
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![17] = The pews
Releases a gas: Corroding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![18] = The elevator (Mechanical)
Releases a gas: Obscuring
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![19] = The window
Releases a gas: Blinding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![20] = The magic circle
Releases a gas: Weakening
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![21] = The arch
Releases a gas: Poisonous
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![22] = The elevator (Magical)
Releases a gas: Weakening
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![23] = The chute
Releases a gas: Sleep Inducing
Infliciting 1d8 points of damage in addition to the other effects listed

Gas Trap![24] = The manacles
Releases a gas: Obscuring
Infliciting 1d8 points of damage in addition to the other effects listed

Gas Trap![25] = The trophy
Releases a gas: Corroding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![26] = The organ
Releases a gas: Corroding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![27] = The arch
Releases a gas: Corroding
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![28] = The catwalk
Releases a gas: Obscuring
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![29] = The jar
Releases a gas: Weakening
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![30] = The crate
Releases a gas: Sleep Inducing
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![31] = The concealed door
Releases a gas: Terrifying
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![32] = The fresco
Releases a gas: Corroding
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![33] = The catwalk
Releases a gas: Blinding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![34] = The bridge
Releases a gas: Corroding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![35] = The basin
Releases a gas: Sleep Inducing
Infliciting 1d8 points of damage in addition to the other effects listed

Gas Trap![36] = The urn
Releases a gas: Corroding
Infliciting 10d6 points of damage in addition to the other effects listed

Gas Trap![37] = The table
Releases a gas: Nauseating
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![38] = The pillar
Releases a gas: Terrifying
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![39] = The barrel
Releases a gas: Nauseating
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![40] = The hourglass
Releases a gas: Sleep Inducing
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![41] = The button/switch
Releases a gas: Terrifying
Infliciting 10d10points of damage in addition to the other effects listed

Gas Trap![42] = The idol
Releases a gas: Corroding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![43] = The chest
Releases a gas: Nauseating
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![44] = The portcullis
Releases a gas: Terrifying
Infliciting 1d4 points of damage in addition to the other effects listed

Gas Trap![45] = The stairway
Releases a gas: Poisonous
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![46] = The manacles
Releases a gas: Blinding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![47] = The aquarium
Releases a gas: Poisonous
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![48] = The idol
Releases a gas: Obscuring
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![49] = The idol
Releases a gas: Poisonous
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![50] = The mosaic
Releases a gas: Terrifying
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![51] = The cask
Releases a gas: Sleep Inducing
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![52] = The furnace
Releases a gas: Terrifying
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![53] = The holy/unholy font
Releases a gas: Terrifying
Infliciting 1d8 points of damage in addition to the other effects listed

Gas Trap![54] = The illusion
Releases a gas: Sleep Inducing
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![55] = The bell
Releases a gas: Obscuring
Infliciting 4d6 points of damage in addition to the other effects listed

Gas Trap![56] = The barrel
Releases a gas: Poisonous
Infliciting 6d6 points of damage in addition to the other effects listed

Gas Trap![57] = The column
Releases a gas: Corroding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![58] = The candelabra
Releases a gas: Poisonous
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![59] = The candelabra
Releases a gas: Obscuring
Infliciting 4d6 points of damage in addition to the other effects listed

Gas Trap![60] = The caldron
Releases a gas: Corroding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![61] = The statue
Releases a gas: Obscuring
Infliciting 1d8 points of damage in addition to the other effects listed

Gas Trap![62] = The pews
Releases a gas: Weakening
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![63] = The rug
Releases a gas: Obscuring
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![64] = The crate
Releases a gas: Sleep Inducing
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![65] = The trunk
Releases a gas: Blinding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![66] = The ladder
Releases a gas: Obscuring
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![67] = The telescope (Mechanical)
Releases a gas: Poisonous
Infliciting 6d6 points of damage in addition to the other effects listed

Gas Trap![68] = The magic circle
Releases a gas: Nauseating
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![69] = The bridge
Releases a gas: Obscuring
Infliciting 4d6 points of damage in addition to the other effects listed

Gas Trap![70] = The furnace
Releases a gas: Poisonous
Infliciting 6d6 points of damage in addition to the other effects listed

Gas Trap![71] = The hourglass
Releases a gas: Sleep Inducing
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![72] = The secret door
Releases a gas: Obscuring
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![73] = The wall
Releases a gas: Blinding
Infliciting 8d6 points of damage in addition to the other effects listed

Gas Trap![74] = The vase
Releases a gas: Terrifying
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![75] = The concealed door
Releases a gas: Blinding
Infliciting 6d6 points of damage in addition to the other effects listed

Gas Trap![76] = The illusion
Releases a gas: Corroding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![77] = The statue
Releases a gas: Blinding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![78] = The basin
Releases a gas: Obscuring
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![79] = The window
Releases a gas: Poisonous
Infliciting 1d4 points of damage in addition to the other effects listed

Gas Trap![80] = The barrel
Releases a gas: Sleep Inducing
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![81] = The brazier
Releases a gas: Blinding
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![82] = The fountain
Releases a gas: Blinding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![83] = The bridge
Releases a gas: Corroding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![84] = The table
Releases a gas: Obscuring
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![85] = The crystal ball
Releases a gas: Weakening
Infliciting 8d6 points of damage in addition to the other effects listed

Gas Trap![86] = The pit
Releases a gas: Blinding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![87] = The railing
Releases a gas: Blinding
Infliciting 10d10points of damage in addition to the other effects listed

Gas Trap![88] = The pool
Releases a gas: Nauseating
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![89] = The armor
Releases a gas: Terrifying
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![90] = The table
Releases a gas: Terrifying
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![91] = The illusion
Releases a gas: Corroding
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![92] = The catwalk
Releases a gas: Nauseating
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![93] = The escalator (Magical)
Releases a gas: Weakening
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![94] = The chandelier
Releases a gas: Nauseating
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![95] = The ceiling
Releases a gas: Obscuring
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![96] = The bed
Releases a gas: Weakening
Infliciting 3d6 points of damage in addition to the other effects listed

Gas Trap![97] = The holy/unholy font
Releases a gas: Poisonous
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![98] = The crystal ball
Releases a gas: Corroding
Infliciting 2d6 points of damage in addition to the other effects listed

Gas Trap![99] = The mirror
Releases a gas: Sleep Inducing
Infliciting 4d6 points of damage in addition to the other effects listed

